File: server/apps/api-service/src/main/java/com/sharetimer/apiservice/application/service/TimerService.java
package com.sharetimer.apiservice.application.service;

import java.time.Instant;
import java.util.UUID;

import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.sharetimer.apiservice.adapter.in.web.dto.TimerCreateRes;
import com.sharetimer.apiservice.adapter.in.web.dto.TimerInfoRes;
import com.sharetimer.apiservice.application.port.in.TimerUseCase;
import com.sharetimer.apiservice.application.port.in.command.AddTimestampCommand;
import com.sharetimer.apiservice.application.port.in.command.CreateTimerCommand;
import com.sharetimer.apiservice.application.port.in.command.DeleteTimerCommand;
import com.sharetimer.apiservice.application.port.in.command.GetTimerCommand;
import com.sharetimer.apiservice.application.port.in.command.UpdateTimerCommand;
import com.sharetimer.apiservice.application.port.out.LoadTimerPort;
import com.sharetimer.apiservice.application.port.out.SaveTimerPort;
import com.sharetimer.apiservice.application.port.out.TimerEventPort;
import com.sharetimer.apiservice.domain.model.Timer;
import com.sharetimer.apiservice.domain.model.Timestamp;
import com.sharetimer.web.support.exception.BadRequestException;
import com.sharetimer.web.support.exception.ForbiddenException;
import com.sharetimer.web.support.exception.NotFoundException;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

@Service
@RequiredArgsConstructor
@Slf4j
public class TimerService implements TimerUseCase {

  private final LoadTimerPort loadTimerPort;
  private final SaveTimerPort saveTimerPort;
  private final TimerEventPort timerEventPort;

  @Override
  @Transactional
  public TimerCreateRes createTimer(CreateTimerCommand command) {
    Timer newTimer = Timer.builder().targetTime(command.targetTime()).ownerToken(UUID.randomUUID()).build();

    Timer savedTimer = saveTimerPort.saveTimer(newTimer);
    log.debug("New timer created: timerId={}, ownerToken={}", savedTimer.getId(),
        savedTimer.getOwnerToken());

    timerEventPort.scheduleExpiration(savedTimer.getId().toString(), savedTimer.getTargetTime());

    return new TimerCreateRes(savedTimer.getId().toString(), savedTimer.getOwnerToken().toString());
  }

  @Override
  @Transactional(readOnly = true)
  @Cacheable(value = "timers", key = "#command.timerId", unless = "#result == null")
  public TimerInfoRes getTimerInfo(GetTimerCommand command) {
    Timer timer = loadTimerPort.loadTimer(UUID.fromString(command.timerId()))
        .orElseThrow(() -> new NotFoundException("TimerNotFound",
            "Timer not found. timerId=" + command.timerId()));

    boolean isOwner = isOwner(command.ownerToken(), timer);

    return TimerInfoRes.from(Instant.now(), isOwner, timer);
  }

  @Override
  @Transactional
  @CacheEvict(value = "timers", key = "#command.timerId")
  public void deleteTimer(DeleteTimerCommand command) {
    saveTimerPort.deleteTimer(UUID.fromString(command.timerId()));
  }

  @Override
  @Transactional
  @CacheEvict(value = "timers", key = "#command.timerId")
  public void updateTimer(UpdateTimerCommand command) {
    Timer timer = loadTimerPort.loadTimer(UUID.fromString(command.timerId()))
        .orElseThrow(() -> new NotFoundException("TimerNotFound",
            "Timer not found. timerId=" + command.timerId()));

    if (!isOwner(command.ownerToken(), timer)) {
      throw new ForbiddenException("OwnerTokenMismatch", "No permission to update timer.");
    }

    if (command.requestTime().isBefore(timer.getUpdatedAt())) {
      log.warn("Old timer update request. timerId={}, requestTime={}, updatedAt={}", command.timerId(),
          command.requestTime(), timer.getUpdatedAt());
      return;
    }

    timer.updateTargetTime(command.targetTime());
    timer.updateUpdatedAt(command.requestTime());

    timerEventPort.scheduleExpiration(
        timer.getId().toString(),
        timer.getTargetTime());
    timerEventPort.publishUpdateTimerTargetTime(
        timer.getId().toString(),
        timer.getUpdatedAt(),
        timer.getTargetTime());
  }

  @Override
  @Transactional
  @CacheEvict(value = "timers", key = "#command.timerId")
  public void addTimestamp(AddTimestampCommand command) {
    Timer timer = loadTimerPort.loadTimer(UUID.fromString(command.timerId()))
        .orElseThrow(() -> new NotFoundException("TimerNotFound",
            "Timer not found. timerId=" + command.timerId()));

    if (command.capturedAt().isAfter(timer.getTargetTime())) {
      throw new BadRequestException("InvalidCapturedAt",
          "Timestamp time is before timer target time. capturedAt=" + command.capturedAt());
    }

    timer.getTimestamps().add(
        Timestamp.builder()
            .timer(timer)
            .targetTime(timer.getTargetTime())
            .capturedAt(command.capturedAt())
            .build());

    timerEventPort.publishAddTimestamp(
        timer.getId().toString(),
        timer.getTargetTime(),
        command.capturedAt());
  }

  private boolean isOwner(String ownerToken, Timer timer) {
    return ownerToken != null && UUID.fromString(ownerToken).equals(timer.getOwnerToken());
  }

}

File: server/apps/api-service/src/main/kotlin/com/sharetimer/apiservice/application/service/TimerService.kt
package com.sharetimer.apiservice.application.service

import com.sharetimer.apiservice.application.port.input.TimerUseCase
import com.sharetimer.apiservice.application.port.input.command.AddTimestampCommand
import com.sharetimer.apiservice.application.port.input.command.CreateTimerCommand
import com.sharetimer.apiservice.application.port.input.command.DeleteTimerCommand
import com.sharetimer.apiservice.application.port.input.command.GetTimerQuery
import com.sharetimer.apiservice.application.port.input.command.UpdateTimerCommand
import com.sharetimer.apiservice.application.port.output.LoadTimerPort
import com.sharetimer.apiservice.application.port.output.SaveTimerPort
import com.sharetimer.apiservice.application.port.output.TimerEventPort
import com.sharetimer.apiservice.domain.model.Timer
import com.sharetimer.web.support.exception.ForbiddenException
import com.sharetimer.web.support.exception.NotFoundException
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.UUID

@Service
class TimerService(
    private val loadTimerPort: LoadTimerPort,
    private val saveTimerPort: SaveTimerPort,
    private val timerEventPort: TimerEventPort,
) : TimerUseCase {
    companion object {
        private val log = LoggerFactory.getLogger(TimerService::class.java)
    }

    @Transactional
    override fun createTimer(cmd: CreateTimerCommand): Timer {
        val newTimer =
            Timer(
                targetTime = cmd.targetTime,
                ownerToken = UUID.randomUUID(),
            )

        val savedTimer = saveTimerPort.saveTimer(newTimer)
        log.debug("New timer created: timerId={}, ownerToken={}", savedTimer.id, savedTimer.ownerToken)

        timerEventPort.scheduleExpiration(savedTimer.id.toString(), savedTimer.targetTime)

        return savedTimer
    }

    @Transactional(readOnly = true)
    override fun getTimerInfo(query: GetTimerQuery): Timer = findTimer(query.timerId)

    @Transactional
    override fun deleteTimer(cmd: DeleteTimerCommand) {
        saveTimerPort.deleteTimer(UUID.fromString(cmd.timerId))
    }

    @Transactional
    override fun updateTimer(cmd: UpdateTimerCommand) {
        val timer = findTimer(cmd.timerId)

        if (!isOwner(cmd.ownerToken, timer)) {
            throw ForbiddenException("OwnerTokenMismatch", "No permission to update timer.")
        }

        runCatching { timer.updateTargetTime(cmd.targetTime, cmd.requestTime) }
            .onFailure {
                log.warn(
                    "Old timer update request. timerId={}, requestTime={}, updatedAt={}",
                    cmd.timerId,
                    cmd.requestTime,
                    timer.updatedAt,
                )
                return
            }

        saveTimerPort.saveTimer(timer)

        timerEventPort.scheduleExpiration(timer.id.toString(), timer.targetTime)
        timerEventPort.publishUpdateTimerTargetTime(
            timer.id.toString(),
            cmd.requestTime,
            timer.targetTime,
        )
    }

    @Transactional
    override fun addTimestamp(cmd: AddTimestampCommand) {
        val timer = findTimer(cmd.timerId)

        timer.addTimestamp(cmd.capturedAt)
        saveTimerPort.saveTimer(timer)

        timerEventPort.publishAddTimestamp(
            timer.id.toString(),
            timer.targetTime,
            cmd.capturedAt,
        )
    }

    private fun findTimer(timerId: String): Timer =
        loadTimerPort.loadTimer(UUID.fromString(timerId))
            ?: throw NotFoundException("TimerNotFound", "Timer not found. timerId=$timerId")

    private fun isOwner(
        ownerToken: String?,
        timer: Timer,
    ): Boolean = ownerToken != null && UUID.fromString(ownerToken) == timer.ownerToken
}
